### PR TITLE
Refactor boilerplate code in Search API

### PR DESCRIPTION
**Pull request by arturalbov**
Refactor boilerplate code in Search API

### CODE REVIEW

File: cassandra-service/src/main/kotlin/fund/cyber/cassandra/common/RepositoriesConfiguration.kt [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    val mappingContext = CassandraMappingContext()
+
+    mappingContext.setInitialEntitySet(CassandraEntityClassScanner.scan(basePackage))
+    mappingContext.setUserTypeResolver(SimpleUserTypeResolver(cluster, keyspace))
+    mappingContext.setCustomConversions(customConversions)
+
+    return mappingContext
+}
+
+fun getKeyspaceSession(cluster: Cluster,
```
Comment (hleb-albau) [inline cassandra-service/src/main/kotlin/fund/cyber/cassandra/common/RepositoriesConfiguration.kt]: format?

File: cassandra-service/src/main/kotlin/fund/cyber/cassandra/common/RepositoriesConfiguration.kt [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+}
+
+/**
+ * Return bean name of search repository by specific chainName
+ *
+ * @param chainName chain name
+ * @return search repository bean name
+ */
+fun Class<*>.searchRepositoryBeanName(chainName: String) = "$chainName$REPOSITORY_NAME_DELIMITER${this.name}"
```
Comment (hleb-albau) [inline cassandra-service/src/main/kotlin/fund/cyber/cassandra/common/RepositoriesConfiguration.kt]: Class<*> ? Is it possible to narrow down  extension scope here?

File: cassandra-service/src/main/kotlin/fund/cyber/cassandra/common/RepositoriesConfiguration.kt [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    private val entityBasePackage: String,
+    private val keyspacePrefix: String
+) : InitializingBean {
+
+    @Autowired
+    private lateinit var applicationContext: GenericApplicationContext
+    @Autowired
+    private lateinit var cluster: Cluster
+
+    abstract fun repositoriesClasses(): List<Class<*>>
```
Comment (hleb-albau) [inline cassandra-service/src/main/kotlin/fund/cyber/cassandra/common/RepositoriesConfiguration.kt]: Class<*> ?

File: search-api/src/main/kotlin/fund/cyber/api/bitcoin/handlers/Block.kt [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    @Bean
+    fun bitcoinBlockItemHandler() = RepositoryItemRequestHandler(
+        "/block/{blockNumber}",
+        BitcoinBlockRepository::class.java
+    ) { request, repository, _ ->
+
+            val blockNumber = request.pathVariable("blockNumber").toLong()
+            val block = repository.findById(blockNumber)
+            ServerResponse.ok().body(block, CqlBitcoinBlock::class.java)
```
Comment (hleb-albau) [inline search-api/src/main/kotlin/fund/cyber/api/bitcoin/handlers/Block.kt]: add switchIfEmpty to all handlers

File: search-api/src/main/kotlin/fund/cyber/api/bitcoin/handlers/Contract.kt [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+@DependsOn("bitcoin-search-repositories")
+class BitcoinContractHandlersConfiguration {
+
+    @Autowired
+    private lateinit var applicationContext: GenericApplicationContext
+
+    @Bean
+    fun bitcoinContractItemHandler() = RepositoryItemRequestHandler(
+        "/contract/{hash}",
+        BitcoinContractSummaryRepository::class.java
```
Comment (hleb-albau) [inline search-api/src/main/kotlin/fund/cyber/api/bitcoin/handlers/Contract.kt]: add ne more class, that takes 2 repos. It's okey.
Context  lookup on each request is not we want.

File: search-api/src/main/kotlin/fund/cyber/api/common/RequestHandler.kt [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+import org.springframework.web.reactive.function.server.HandlerFunction
+import org.springframework.web.reactive.function.server.RequestPredicates
+import org.springframework.web.reactive.function.server.RouterFunction
+import org.springframework.web.reactive.function.server.RouterFunctions
+import org.springframework.web.reactive.function.server.ServerRequest
+import org.springframework.web.reactive.function.server.ServerResponse
+import reactor.core.publisher.Mono
+
+
+class RepositoryItemRequestHandler<T>(
```
Comment (hleb-albau) [inline search-api/src/main/kotlin/fund/cyber/api/common/RequestHandler.kt]: nice!

File: search-api/src/main/kotlin/fund/cyber/api/ethereum/handlers/Contract.kt [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+        val hash = request.pathVariable("hash")
+        val page = request.queryParam("page").orElse("0").toInt()
+        val pageSize = request.queryParam("pageSize").orElse("20").toInt()
+
+
+        var slice = repository
+            .findAllByMinerContractHash(hash.toSearchHashFormat(), CassandraPageRequest.first(pageSize))
+
+        for (i in 1..page) {
```
Comment (hleb-albau) [inline search-api/src/main/kotlin/fund/cyber/api/ethereum/handlers/Contract.kt]: maybe create common function for slice logic?

File: search-api/src/main/kotlin/fund/cyber/api/bitcoin/handlers/Block.kt (hunk 4 above)
Comment (arturalbov) [reply search-api/src/main/kotlin/fund/cyber/api/bitcoin/handlers/Block.kt]: It's out of the task scope. Will be implemented in #178

File: search-api/src/main/kotlin/fund/cyber/api/ethereum/handlers/Contract.kt (hunk 7 above)
Comment (arturalbov) [reply search-api/src/main/kotlin/fund/cyber/api/ethereum/handlers/Contract.kt]: good idea

**Comment by arturalbov** [comment]
Closes #114

**Pull request merged**